decidim/decidim
# Fix pipeline asset absolute URLs

#### :tophat: What? Why?
This is continuation for #9550.

When pipeline assets were referred from `Decidim::AvatarUploader`, they did not return the full URLs but paths to these images. If these images are used e.g. in emails, this would cause similar issues as described at #9550 and #9596 but for the images.

This introduces a new `Decidim::AssetRouter::Pipeline` class which will format these URLs correctly for any context.

At the same time I also renamed the previous asset router to `Decidim::AssetRouter::Storage` to separate these two cases from each other. This class should match the original `Decidim::AssetRouter` apart from some minor comment formatting that I did.

#### :pushpin: Related Issues
- Related to #9550, #9596

#### Testing
You can test this from the console:
- Run `bundle exec rails c` in the development app
- Run `Decidim::UserPresenter.new(Decidim::User.first).avatar_url`
- You should see the full URL to be printed to the console (make sure you have not uploaded an avatar image for that user to get the default URL, otherwise use some other user)
- Note that the URL will not have the protocol in it unless you are running Decidim with the `force_ssl` option
  * You can try with this option by starting the console with `DECIDIM_FORCE_SSL=true bundle exec rails c`
  * If it didn't work make sure to restart spring before starting the console by running `spring stop` before the console command
